Update to readme and added application notes #168

mofosyne opened the pull request:
Issue Ticket: https://github.com/Mozilla-Ocho/llamafile/issues/168

Added recommended path convention for installation as well as application notes.

This commit is based on jart recommendation regarding llamafile convention. 
This is her quote that this is based on:

> I want to enable people to integrate with llamafile any way they like.
> In terms of recommendations and guidance, I've been following
> TheBloke's naming convention when publishing llamafiles to Hugging
> Face https://huggingface.co/jartine I also always use the llamafile
> tag. So what I'd recommend applications do, is iterate all the files
>  tagged llamafile on Hugging Face to present those as choices to the
>  user for LLMs. Be sure to display which user is publishing them, and
> sort by heart count. Then, when you download them, feel free to put
> them in ~/.llamafile. Then, to show the users which models are
> installed, you just look for ~/.llamafile/*.llamafile.

jart commented on APPLICATION.md:
```diff
@@ -0,0 +1,42 @@
+
+While we do not have a package manager for llamafiles, applications developers are recommended
+to search for AI models tagged as `llamafile` in Hugging Face AI repository.
+Be sure to display the publishing user or organisation and to sort by heart count.
```
I would revise my early suggestion to "sort by trending" so we don't end up with permanent victors, and people have a better chance of being able to talk online with their friends about what's popular at any given moment.

jart commented on APPLICATION.md:
```diff
@@ -0,0 +1,42 @@
+
+## Packaging A Llamafile
+
+The file naming convention for each llamafile is `<model-name>.<Quantisation Level>.llamafile` (e.g. `phi-2.Q2_K.llamafile`).
```
I'd suggest to include the model version number, but _only_ if the model has multiple versions, e.g. Mistral 7b. We'd then assume the unversioned one was the first version, and/or check the model card.

jart commented on APPLICATION.md:
```diff
@@ -0,0 +1,42 @@
+Llamafile are completely standalone and portable so does not require installation.
+However we have a path convention for ease of discovery by local applications scripts/program.
+
+- **Linux** : `~/.llamafile/*.llamafile`
```
We use that path for Windows, MacOS, and BSD too.

We may want to suggest a system-wide path. In that case, I'd suggest `/usr/share/llamafile`, `/opt/llamafile`, and `C:\llamafile` be checked. Then maybe suggest `$LLAMAFILE_PATH` be a single (not a list) path that be checked.

mofosyne commented:
Okay revised the readmes based on your suggestion. Also spent some time studying how models naming convention is currenting working in the field and how it's defined in llama.cpp . There is likely issues with the "Llamafile Naming Convention" section but everything else should hopefully be addressed now.

mofosyne commented:
If we settle on `<Model>-<Version>-<Parameters>-<Quantization>.llamafile` we may want to adjust the file creation process to enforce this. Maybe have a quiz? Maybe extract as much as possible from the GGUF file format metadata?

At least according to https://github.com/ggerganov/ggml/blob/master/docs/gguf.md you can get the ggml_type (e.g. `Q6_K` or `F32`), but according to the file format... you can have multiple different mix of tensor type?

mofosyne commented:
Ballob founder of home assistant on what he would require LLM container to do


> I would love to see a standardized API for local LLMs that is not just a 1:1 copying the ChatGPT API. For example, as Home Assistant talks to a random model, we should be able to query that model to see what the model is capable off.

Is this achievable by adding Key Values to the GGML?

> I want to see local LLMs with support for a feature similar or equivalent to OpenAI functions. We cannot include all possible information in the prompt and we need to allow LLMs to make actions to be useful. Constrained grammars do look like an possible alternative. Creating a prompt to write JSON is possible but need quite an elaborate prompt and even then the LLM can make errors. We want to make sure that all JSON coming out of the model is directly actionable without having to ask the LLM what they might have meant for a specific value.

Having a recommended way to easily constraint output to Json would help in the application notes.

vivianlocke643333 commented:
If you have lost your funds and are looking for a way to recover them, there are legitimate and genuine recovery firms that you can contact. Many people have found themselves in similar situations, and some have been able to recover their funds through these recovery companies. However, for those who are not aware of such recovery firms, they may have lost their entire investment. One reputable recovery company that has been assisting people in recovering their wealth without upfront fees is CRYPTO RECOVERY FIXED COMPANY. You can visit their website at @https://cryptorecoveryfixed.com or send them an email at @Cryptorecoveryfixed@gmail.com to get assistance with your recovery.

mofosyne commented:
Just did a rebase to keep this PR up to date with main

mofosyne commented:
While rebasing https://github.com/ggerganov/llama.cpp/pull/4858 , decided to review my naming convention proposal and noticed that mixtral has a new naming approach for their model like `8x7B` to indicate 8 experts 7B quant.

I've added the new addition to both the llama.cpp default filename PR and also updated the readme notes in this repo's PR as well.

mofosyne commented:
https://github.com/ggerganov/llama.cpp/issues/7165 now merged in so `<Model>-<Version>-<ExpertsCount>x<Parameters>-<Quantization>.gguf` is now more canonical.

mofosyne commented:
rebase to be on top of latest changes and squash all the other fixup commits. Did another review to make sure the doc matches with the now merged in change to llama.cpp convert.py

mofosyne commented:
On a side note what generates `<!-- README_llamafile.md-provided-files start -->`  that I see occasionally?
